File: multiple-languages/go/alicloudroscdkoss/IBucketWorm.go
package alicloudroscdkoss

import (
	_jsii_ "github.com/aws/jsii-runtime-go/runtime"

	"github.com/alibabacloud-go/ros-cdk/alicloudroscdkcore"
	"github.com/alibabacloud-go/ros-cdk/alicloudroscdkoss/internal"
)

// Represents a `BucketWorm`.
type IBucketWorm interface {
	alicloudroscdkcore.IResource
	// Attribute CreateTime: The creation time of the resource.
	AttrCreateTime() interface{}
	// Attribute ExpirationDate: Retention Policy Expiration Time.
	AttrExpirationDate() interface{}
	// Attribute RetentionPeriodInDays: The specified number of days to retain the Object.
	AttrRetentionPeriodInDays() interface{}
	// Attribute WormId: The ID of the retention policy.
	AttrWormId() interface{}
	Props() *BucketWormProps
}

// The jsii proxy for IBucketWorm
type jsiiProxy_IBucketWorm struct {
	internal.Type__alicloudroscdkcoreIResource
}

func (j *jsiiProxy_IBucketWorm) AttrCreateTime() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrCreateTime",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IBucketWorm) AttrExpirationDate() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrExpirationDate",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IBucketWorm) AttrRetentionPeriodInDays() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrRetentionPeriodInDays",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IBucketWorm) AttrWormId() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrWormId",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IBucketWorm) Props() *BucketWormProps {
	var returns *BucketWormProps
	_jsii_.Get(
		j,
		"props",
		&returns,
	)
	return returns
}

File: multiple-languages/dotnet/AlibabaCloud.SDK.ROS.CDK.Oss/AlibabaCloud/SDK/ROS/CDK/Oss/IBucketWorm.cs
using Amazon.JSII.Runtime.Deputy;

#pragma warning disable CS0672,CS0809,CS1591

namespace AlibabaCloud.SDK.ROS.CDK.Oss
{
    /// <summary>Represents a `BucketWorm`.</summary>
    [JsiiInterface(nativeType: typeof(IBucketWorm), fullyQualifiedName: "@alicloud/ros-cdk-oss.IBucketWorm")]
    public interface IBucketWorm : AlibabaCloud.SDK.ROS.CDK.Core.IResource
    {
        /// <summary>Attribute CreateTime: The creation time of the resource.</summary>
        [JsiiProperty(name: "attrCreateTime", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrCreateTime
        {
            get;
        }

        /// <summary>Attribute ExpirationDate: Retention Policy Expiration Time.</summary>
        [JsiiProperty(name: "attrExpirationDate", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrExpirationDate
        {
            get;
        }

        /// <summary>Attribute RetentionPeriodInDays: The specified number of days to retain the Object.</summary>
        [JsiiProperty(name: "attrRetentionPeriodInDays", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrRetentionPeriodInDays
        {
            get;
        }

        /// <summary>Attribute WormId: The ID of the retention policy.</summary>
        [JsiiProperty(name: "attrWormId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrWormId
        {
            get;
        }

        [JsiiProperty(name: "props", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-oss.BucketWormProps\"}")]
        AlibabaCloud.SDK.ROS.CDK.Oss.IBucketWormProps Props
        {
            get;
        }

        /// <summary>Represents a `BucketWorm`.</summary>
        [JsiiTypeProxy(nativeType: typeof(IBucketWorm), fullyQualifiedName: "@alicloud/ros-cdk-oss.IBucketWorm")]
        internal sealed class _Proxy : DeputyBase, AlibabaCloud.SDK.ROS.CDK.Oss.IBucketWorm
        {
            private _Proxy(ByRefValue reference): base(reference)
            {
            }

            /// <summary>Attribute CreateTime: The creation time of the resource.</summary>
            [JsiiProperty(name: "attrCreateTime", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrCreateTime
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute ExpirationDate: Retention Policy Expiration Time.</summary>
            [JsiiProperty(name: "attrExpirationDate", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrExpirationDate
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute RetentionPeriodInDays: The specified number of days to retain the Object.</summary>
            [JsiiProperty(name: "attrRetentionPeriodInDays", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrRetentionPeriodInDays
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute WormId: The ID of the retention policy.</summary>
            [JsiiProperty(name: "attrWormId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrWormId
            {
                get => GetInstanceProperty<object>()!;
            }

            [JsiiProperty(name: "props", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-oss.BucketWormProps\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Oss.IBucketWormProps Props
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Oss.IBucketWormProps>()!;
            }

            /// <summary>The environment this resource belongs to.</summary>
            /// <remarks>
            /// For resources that are created and managed by the CDK
            /// (generally, those created by creating new class instances like Role, Bucket, etc.),
            /// this is always the same as the environment of the stack they belong to;
            /// however, for imported resources
            /// (those obtained from static methods like fromRoleArn, fromBucketName, etc.),
            /// that might be different than the stack they were imported into.
            /// </remarks>
            [JsiiProperty(name: "env", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-core.IResourceEnvironment\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Core.IResourceEnvironment Env
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Core.IResourceEnvironment>()!;
            }

            /// <summary>The stack in which this resource is defined.</summary>
            [JsiiProperty(name: "stack", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-core.Stack\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Core.Stack Stack
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Core.Stack>()!;
            }

            /// <summary>The construct tree node for this construct.</summary>
            [JsiiProperty(name: "node", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-core.ConstructNode\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Core.ConstructNode Node
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Core.ConstructNode>()!;
            }
        }
    }
}
